compiler/optimizer/OSRDefAnalysis.cpp

```diff
@@ -1378,6 +1408,82 @@ int32_t TR_OSRExceptionEdgeRemoval::perform()
+   if (!disableDeadStoreCleanup)
+      {
+      TR_BitVector alwaysDead(comp()->trMemory()->currentStackRegion());
+      TR_Stack<TR::Block *> osrBlocks(trMemory(), 8, false, stackAlloc);
+      _seenDeadStores = new (trStackMemory()) TR_BitVector(comp()->trMemory()->currentStackRegion());
+
+      // OSR induce and code blocks will contain dead stores for the next
+      // frame.
+      for (cfgNode = cfg->getFirstNode(); cfgNode; cfgNode = cfgNode->getNext())
+         {
+         TR::Block *block = toBlock(cfgNode);
+         if (!block->isOSRCodeBlock())
+            continue;
+
+         // Find the prepareForOSR call
+         TR::TreeTop *tt = block->getFirstRealTreeTop();
+         while (tt && !(tt->getNode()->getNumChildren() > 0 && tt->getNode()->getFirstChild()->getOpCode().isCall()))
+            tt = tt->getNextTreeTop();
+         TR_ASSERT(tt && tt->getNode()->getFirstChild()->getOpCode().isCall(), "OSRCode blocks must contain a call to prepareForOSR and no other calls");
+         TR::Node *prepareForOSR = tt->getNode()->getFirstChild();
+
+         // Collect symrefs that are always dead
+         alwaysDead.empty();
+         bool firstPass = true;
+         for (auto edge = block->getPredecessors().begin(); edge != block->getPredecessors().end(); ++edge)
+            {
+            TR::Block *osrBlock = toBlock((*edge)->getFrom());
+            TR_ASSERT(osrBlock && (osrBlock->isOSRCodeBlock() || osrBlock->isOSRCatchBlock()), "Predecessors to an OSR code block should be OSR code or catch blocks");
+            if (osrBlock->isOSRCodeBlock() && addDeadStores(osrBlock, alwaysDead, !firstPass))
+               osrBlocks.push(osrBlock);
+            else if (osrBlock->isOSRCatchBlock())
+               {
+               for (auto exEdge = osrBlock->getExceptionPredecessors().begin(); exEdge != osrBlock->getExceptionPredecessors().end(); ++exEdge)
+                  {
+                  TR::Block *exBlock = toBlock((*exEdge)->getFrom());
+                  TR_ASSERT(exBlock && exBlock->isOSRInduceBlock(), "Predecessors to an OSR catch block should be OSR induce blocks");
+                  if (addDeadStores(exBlock, alwaysDead, !firstPass))
+                     osrBlocks.push(exBlock);
+                  firstPass = false;
+                  }
+               }
+            firstPass = false;
+            }
+
+         // Remove loads that are always dead
+         for (int32_t i = 0 ; i < prepareForOSR->getNumChildren(); ++i)
+            {
+            TR::Node *child = prepareForOSR->getChild(i);
+            if (child->getOpCode().isLoadVar() && alwaysDead.get(child->getSymbolReference()->getReferenceNumber()))
+               {
+               if (performTransformation(comp(), "%s: Remove dead OSR load [%p] from OSR code block_%d\n", OPT_DETAILS, child, block->getNumber()))
+                  {
+                  child->decReferenceCount();
```

Comment:
Shouldn't we just recreate the node in place because the node has an identical value by the nature of the transform and so any use will benefit from knowing it is zero...

Reply:
A recreate to the corresponding type has been added. I could not find a similar example of this, but I believe this is correct.